serveradmin/serverdb/views.py

```diff
@@ -87,38 +100,35 @@ def history(request):
     if not object_id:
         raise Http404
 
-    adds = ChangeAdd.objects.filter(server_id=object_id)
-    updates = ChangeUpdate.objects.filter(server_id=object_id)
-    deletes = ChangeDelete.objects.filter(server_id=object_id).select_related()
+    # TODO: Change data model so that server_id is moved to ChangeCommit.
+    #       To me this would make most sense. If that is not possible (why?!)
+    #       write a proper (custom) query that performs.
+    where = {'server_id': object_id}
+    adds = ChangeAdd.objects.filter(**where)
+    updates = ChangeUpdate.objects.filter(**where)
+    deletes = ChangeDelete.objects.filter(**where)
 
     # @TODO Transform to PostgreSQL JSON and put a index here
     if search_string:
         adds = adds.filter(attributes_json__contains=search_string)
         updates = updates.filter(updates_json__contains=search_string)
         deletes = deletes.filter(attributes_json__contains=search_string)
 
-    adds = adds.select_related()
-    updates = updates.select_related()
-
     if commit_id:
         adds = adds.filter(commit__pk=commit_id)
         updates = updates.filter(commit__pk=commit_id)
         deletes = deletes.filter(commit__pk=commit_id)
 
+    related = ('commit__app', 'commit__user')
+    adds = adds.select_related(*related)
```

Comment:
I am not familiar with the django models here. Maybe it won't change anything, but before there was the `select_related()` call, which was actually doing something. If it is still necessary, I guess it should be changed to:
``` python
adds = adds.select_related().select_related(*related)
```

Reply:
i don't think this would work. `select_related` returns related objects. in this case `commit`/`ChangeCommit`. but i also want to "prefetch" commit's relations, too, to save on many unnecessary queries. because the template will try to access `app` and `user` of `commit` and this will trigger twice as many separate queries as there are commit rows. by passing those related fields explicitly, you can optimize it to query everything at once

Reply:
I understand that you want to add the related fields from commit, too, and that it is needed. My comment was not removing that. My comment was stating, that by removing the `select_related()` call, we have a behavior change as it won't select other relations automatically anymore. From django:
```
In these cases it is possible to call select_related() with no arguments. This will follow all non-null foreign keys it can find - nullable foreign keys must be specified.
```
It is anyway not recommended and might not be necessary anyway (as I said, I am not familiar with the models here), just want to point out that it might be a behavior change and if we still need it, it should be:
``` python
adds = adds.select_related().select_related(*related)
```
If we don't need the other relations, which would be selected by `select_related()`, then ignore that comment. ^^

Reply:
i got what you meant. what i mean is that chaining those calls does not have any effect :D if you first select all related and afterwards only a subset of it, then it will choose the subset only anyway

Reply:
After call it is now clear. `select_related()` is not creating a list but just setting a property to `True`, which will be overwritten by the next explicit `select_related(...)` call. The documentation is a bit misleading there, as it sounds like that the `select_related()` call will create a list directly and not just set a property to `True`.
So, the chaining would not make sense.